core/src/main/java/org/infinispan/transaction/xa/XaTransactionTable.java

```diff
@@ -153,23 +172,23 @@ void end(LocalXaTransaction localTransaction) {
          log.tracef("end called on tx %s(%s)", localTransaction.getGlobalTransaction(), cacheName);
    }
 
-   void forget(Xid externalXid) throws XAException {
+   CompletionStage<Void> forget(Xid externalXid) {
       Xid xid = convertXid(externalXid);
       if (trace)
          log.tracef("forget called for xid %s", xid);
-      try {
-         if (isRecoveryEnabled()) {
-            recoveryManager.removeRecoveryInformation(null, xid, true, null, false);
-         } else {
-            if (trace)
-               log.trace("Recovery not enabled");
-         }
-      } catch (Exception e) {
-         log.warnExceptionRemovingRecovery(e);
-         XAException xe = new XAException(XAException.XAER_RMERR);
-         xe.initCause(e);
-         throw xe;
+      if (isRecoveryEnabled()) {
+         return recoveryManager.removeRecoveryInformation(null, xid, null, false)
+               .exceptionally(t -> {
+                  log.warnExceptionRemovingRecovery(t);
+                  XAException xe = new XAException(XAException.XAER_RMERR);
+                  xe.initCause(t);
+                  throw CompletableFutures.asCompletionException(xe);
```

Comment:
Since we know 100% `xe` is not a `CompletionException`, I would find `throw new CompletionException(xe)` clearer.

Reply:
Okay.